Make UUri specification more formal (#121)
eclipse-uprotocol/up-spec: basics/uri.adoc

under the License.

SPDX-FileType: DOCUMENTATION
SPDX-License-Identifier: Apache-2.0
----

== Overview 

uProtocol uses https://datatracker.ietf.org/doc/html/rfc3986[Uniform Resource Identifiers] (URI) to uniquely identify (and/or address) resources such as devices, software components, topics, methods, etc on a computer network. https://datatracker.ietf.org/doc/html/rfc3986[RFC3986] defines the structure of a URI as follows:

.URI
image::rfc3986.png[RFC3986 URI]

Using this terminology, uProtocol URIs (UUri) can be structured like this:

.UUri Specification
[#uuri-specification]

> image::uri.drawio.svg[uProtocol Uri Specification]

Author: I'll do my best :-)

A UUri can be represented as

> * an instantiation of the <<_data_model,UUri Data Model>> in a supported programming language,

Author: I'd like to refer to the section, not the UML class diagram here ...

> * an instantiation of the <<_data_model,UUri Data Model>> in a supported programming language,

Reviewer: It should refer to the section with that title unless you named the title and the diagram the same. The reason why I proposed the change is that the link did not work for me when I rendered it in the PR.

> * a URI containing (logical) names only as defined by <<_long_form>>,

Author: is there a particular reason why you would rather use the section title instead of the identifier created by adoc for it?

> * a URI containing (logical) names only as defined by <<_long_form>>,

Reviewer: Links were not working for me (see previous comment). If it works for you in github then we can resolve these changes

> * a URI containing (logical) names only as defined by <<_long_form>>,

Author: Ah, I see what you mean. The links work in the vscode preview :-( I will try to fix this ...

* a URI containing numeric identifiers as defined by <<_short_form>> or
* a byte array as defined by <<_micro_form>>.

In the remainder of this document, the term _UUri_ is used to refer to an instantiation of the data model, whereas the terms _long form URI_, _short form URI_ and _micro form URI_ are used to refer to its other representations respectively.

== Data Model

The UUri data model is defined in link:../up-core-api/uprotocol/uri.proto[uri.proto] using the proto3 type system.
The diagram below provides an overview of the types and their relations using UML2 notation.

.UUri Data Model
[#uuri-data-model]

> [plantuml, format="png", width="100%", id="UUri Object Model"]

Author: Ok, I will replace this with a Mermaid class diagram

----
@startuml
class UAuthority {
  name: string [0..1]
}
class UEntity {
  name: string [0..1]
  id: uint32 [0..1]
  version_major: uint32 [0..1]
  version_minor: uint32 [0..1]
}
class UResource {
  name: string [0..1]
  instance: string [0..1]
  message: string [0..1]
  id: uint32 [0..1]
}
class UUri

hide empty members

UAuthority "authority" <--* UUri
UEntity "entity" <--* UUri
UResource "resource" <--* UUri
@enduml
----

> NOTE: All properties of UAuthority, UEntity and UResource are _optional_, whereas all of UUri's aggregation relations have cardinality 1, i.e. the `authority`, `entity` and `resource` properties are _never_ empty.

Author: In the proto3 file, all of UUri's properties are optional, i.e. they can be empty. In the class model, UUri's properties (represented by its aggregations) are **not** optional because they have cardinality 1

----
--

.REQ~RESOURCE-MESSAGE
[#req~resource-message~1]
--
The `message` property *MUST* either be empty or contain at least one character.

[source,ocl]
----
context UResource
inv: message->NotEmpty() implies message.length() > 0
----
--

.REQ~RESOURCE-ID
[#req~resource-id~1]
--
The `id` property *MUST* either be empty or contain a value from interval `[0,0xFFFF]`.

[source]
----
context UResource
inv: id->notEmpty() implies id <= 0xFFFF
----
--

.REQ~RESOURCE-RPC-METHOD
[#req~resource-rpc-method~1]
--
A UResource *MAY* be used to identify an RPC service method if its `isRpcMethod` predicate as defined below evaluates to `true`.

[source,ocl]
----
context UResource
def: isRpcMethod() : Boolean =
  ( name = "rpc" and instance->notEmpty() )
  or
  ( id > 0 and id < 0x8000 )
----
--

.REQ~RESOURCE-RPC-RESPONSE
[#req~resource-rpc-response~1]
--
A UResource *MAY* be used to identify an RPC response address if its `isRpcResponse` predicate as defined below evaluates to `true`.

[source,ocl]
----
context UResource
def: isRpcResponse() : Boolean =
  ( name = "rpc" and instance = "response" )
  or
  id = 0 
----
--

=== UUri

.REQ~UURI-RESOLVED
[#req~uuri-resolved~1]
--
[sidebar]

> Is this really needed? It is unclear where this check might be needed.

Author: Does this mean that we can **only** serialize to short form, if UUri has a non-empty entity.id and resource.id? If so, that would not match the other direction, i.e. we currently are able to deserialize a short form URI that does not contain an entity.id and/or resource.id (-> wildcards) ...

> A UUri which contains corresponding numerical identifiers for logical names used in its components, is called a _resolved_ UUri.

Reviewer: A UUri that contains names and numbers is resolved

[source,ocl]
----
context UUri
def: isResolved() : Boolean =
  ( entity.name->notEmpty() implies entity.id->notEmpty() )
  and
  ( resource.name->notEmpty() implies resource.id.notEmpty() )
----

> NOTE: This means that a UUri which _only_ contains numerical identifiers is a _resolved_ UUri.

Reviewer: no it needs names and numbers

Each property value *MUST* be mapped to a _segment_ following the rules defined in link:https://datatracker.ietf.org/doc/html/rfc3986#section-3.3[RFC3986, Section 3.3]. An empty property *MUST* be mapped to an empty segment. 

The `entity.version_major` *MUST* be mapped to the decimal string representation of the property value, e.g. a (hex) value of `0x1a` is mapped to the (decimal) string literal `26`.

The `resource.name` and `resource.instance` *MUST* be mapped to a single _segment_ using the following production:

[source,abnf]
----
resource = [ resource.name [ '.' resource.instance ]]
----

NOTE The `resource` _segment_ will be empty if `resource.name` is empty.
--

.REQ~LONG-FORM-FRAGMENT
[#req~long-form-fragment~1]
--
A long form URI's _fragment_ *MUST* be empty, if any of the UUri's `resource.name` or `resource.message` are empty.

Otherwise, a long form URI's _fragment_ *MUST* be mapped to/from the UUri's `resource.message` property following the rules defined in https://datatracker.ietf.org/doc/html/rfc3986#section-3.5[RFC3986, Section 3.5].
--

.REQ~LONG_FORM_VALIDATION
[#req~long-form-validation~1]
--
[sidebar]

> Is this needed? It currently only verifies that the invariants hold (which, by definition, they should do anyway) ...

Reviewer: If you build a UUri from Short form string then this would be false as there is no name for entity or resource

Author: Is this about the following?
1. Short form URI `up://my-vin/1234/1/5678` gets deserialized into a UUri
2. UUri shall now be serialized to a long form URI but that would result in an empty URI because the entity.id and resource.id may not have been resolved to their corresponding names

I see your point but it (again) raises the question if we allow empty properties during serialization or not. The latter would mean that we do not support generating URIs that contain wildcards and that serialization behavior is inconsistent with deserialization, where we (want to?) support wilcards (i.e. empty path segments) ...

--

The remainder of this section defines the requirements for the data model by means of _invariants_ which must hold true at all times. Additionally, there are some _predicates_ that can be used by uEntities to determine if a UUri contains the information they are interested in. Both the invariants and the predicates are (formally) defined by means of link:https://www.omg.org/spec/OCL/2.4[Object Constraint Language] (OCL) expressions on the UUri data model. The link:https://modeling-languages.com/ocl-tutorial/[OCL Tutorial] provides an overview of the basic concepts used in the definitions.

=== Authority

An authority represents the deployment location of a specific software entity. The location is represented by means of a logical identifier like a domain name (e.g. `mcu1.example.com`) or a Vehicle Identification Number (VIN).

[.specitem,oft-sid="dsn~authority-name-length~1",oft-needs="impl,utest"]
--
A UUri's `authority_name` *MUST NOT* exceed 128 characters in length.

[source,ocl]
----

> inv: authority_name.length() <= 128

Reviewer: Where did this requirement come from?

Author: We had been discussing that we wanted to limit the length of the authority name to some degree. I just picked 128 as a first shot. Do you still think that we should limit the length?

----
--

=== uEntity

A uProtocol software entity (uEntity) is a piece of software deployed somewhere on a network host.
uEntities are uniquely identified within a system by means of the type and version of the service interface that they implement and a numerical identifier which denotes the particular instance of the service (type) in case multiple instances are available on the system.

NOTE: uEntities which produce events for consumption by other uEntities assume a _Service_ role. +
uEntities which consume events produced by other uEntities assume an _Application_ role. +
uEntities *MAY* assume both the Service and Application roles.

[.specitem,oft-sid="dsn~entity-id~1"]
--
A UUri's `ue_id` property value determines the type and instance of service being referred to. +

> The value's _most_ significant 16 bits represent the _service instance_. +

Reviewer: We should put a note that a service instance of 0 means "don't care".

Author: I have added that to the section on pattern matching ...

> The value's _least_ significant 16 bits represent the _service type_.

Reviewer: identifier and not type

Author: We usually talk about _instances_ of a certain _type_, don't we? A service _identifier_ to me suggests that we are talking about a particular instance already, instead of its type.

   *                     constraint(s).
   */
  public final void validateRpcMethod() throws UuriValidationException {
    if (!this.isRpcMethod()) {
      throw new UuriValidationException("not an RPC Method URI");
    }
  }
}
----

Similarly, a Rust library might implement a `UriValidator` struct providing corresponding functions

[source,rust]
----
pub struct UriValidator {}

impl UriValidator {
  pub fn is_rpc_method(uri: &UUri) -> bool {
    ...
  }
  pub fn validate_rpc_method(uri: &UUri) -> Result<(), UuriValidationError> {
    if is_rpc_method(uri) {
      Ok(())
    } else {
      Err(UuriValidationError::new("not an RPC Method URI"))
    }
  }
}
----

or implement the functions on the `UUri` struct

[source,rust]
----
impl UUri {
  pub fn is_rpc_method(&self) -> bool {
    ...
  }
  pub fn validate_rpc_method(&self) -> Result<(), UuriValidationError> {
    if self.is_rpc_method() {
      Ok(())
    } else {
      Err(UuriValidationError::new("not an RPC Method URI"))
    }
  }
}
----

== Pattern Matching

A UUri can be used to define a _pattern_ that other UUris can then be matched against.

[.specitem,oft-sid="dsn~pattern-matching~1",oft-needs="impl,utest"]
--

> A UUri that is used as a pattern to match other UUris against

Reviewer: This is duplicated from the sentence above

* *MAY* have its `authority_name` set to the `*` (`U+002A`, Asterisk) character in order to match _any_ (including _no_) authority,

> * *MAY* have the _service type_ part of its `ue_id` set to `0xFFFF` in order to match _any_ type,

Reviewer: maybe we need to call it service number to differentiate the instance id?

Author: not sure what you mean, replace _service type_ with _service number_ or replace _service instance_ with _service number_?

Reviewer: Re-reading some/ip spec for a little inspiration, they have service_id and service_instance_id, no where do we refer to service_id as type hence why I wanted to keep the id or number but not type.

Author: then let's stick to _service id_ and service instance id_. I'll update accordingly ...

> * *MAY* have the _service instance_ part of its `ue_id` set to `0x0000` in order to match _any_ instance,

Reviewer: I like this!! better than what I was thinking and coding in Java...

+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
----

[%autowidth]
|===
|Field |Description

|UP_VERSION
| The uProtocol (major) version being used. Occupies bits 0 through 7 (octet 0). +
At the time of writing, *MUST* be set to `0x01`.

|TYPE
a| The type of authority contained in this micro URI. Occupies bits 8 through 15 (octet 1). +
*MUST* be set to `0x00` for a local URI.

|URESOURCE_ID
a|The identifier of the URI's resource. Occupies bits 16 through 31 (octets 2-3). +
*MUST* contain the value of the UUri's `resource.id` property.

|UENTITY_ID
a|The identifier of the URI's entity. Occupies bits 32 through 47 (octets 4-5). +
*MUST* contain the value of the UUri's `entity.id` property.

|UE_VERSION
a|The major version of the URI's entity. Occupies bits 48 through 55 (octet 6). +
*MUST* contain the value of the UUri's `entity.version_major` property.

|UNUSED
|Padding to the next 32bit word. Occupies bits 56 through 63 (octet 7). +
*MUST* be set to `0x00`.

|===
--

=== Remote Micro Uri

.REQ~MICRO-FORM-REMOTE
[#req~micro-form-remote~1]
--

> A UUri that contains a non-empty authority *MUST* be mapped to a byte array in network byte order as defined below. The structure extends the one defined by <<_local_micro_uri>>.

Reviewer: ```suggestion
A UUri that contains a non-empty authority *MUST* be mapped to a byte array in network byte order as defined below. The structure extends the one defined by <<Local Micro Uri>>.
```

+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|        UENTITY_ID             |  UE_VERSION   |   UNUSED      |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|   NAME_LEN    |        UAUTHORITY_NAME  ...                   |
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
----

[%autowidth]
|===
|Field |Description

|TYPE
a| The type of authority contained in this micro URI. Occupies bits 8 through 15 (octet 1). +
*MUST* be set to `0x03` for a URI with an authority.

|NAME_LEN
a|The length of the `UAUTHORITY_NAME`. Occupies bits 64 through 71 (octet 24). +
*MUST* contain the length of the `UAUTHORITY_NAME` field in bytes.

> |UAUTHORITY_NAME

Reviewer: @sophokles73, This is simpler in that there is only local vs remote however it means that a 4 byte IPv4 address could take up to 15 bytes and the IPv6 16 bytes could take up to 39 bytes.

Author: I just wanted to point out that you **can** also use IP address strings for the name. But our overall idea was to suggest that uEntities use logical names only for the authority ...

a|The URI's authority. Occupies bits 72 through ( 72 + 8*NAME_LEN - 1 ). +
*MUST* contain the UTF8 encoding of the UUri's `authority.name` property.

|===

NOTE: The authority is not padded to the next 32 bit word.
--

=== Examples

The examples below can be used as test vectors for implementing client libraries.

[%autowidth]
|===
|authority |entity |version |resource |Short form URI

| none
| `0x0000`
| `0x03`
| `0x8000`
l|[
  0x01, // version
  0x00, // type
  0x80, 0x00, // UResource ID
  0x00, 0x00, // UEntity ID
  0x03, // UEntity major version
  0x00, // unused
]

> l|`1G1YZ23J9P5800001`

Reviewer: ```suggestion
l|1G1YZ23J9P5800001
```

Author: hmm, the VIN also is a literal value. We format the hex values accordingly so, for consistency, I'd rather keep this as is. Alternatively we can remove the formatting for all values ...

[source,java]
----
public final class UriSerializer {
  /**
   * @returns The URI parsed from the given string representation.
   * @throws UuriSerializationException if the given string is not a long form
   *                     UUri representation. The exception may contain details
   *                     regarding the violated constraint(s).
   */
  public static UUri deserializeLongFormUri(String: uri) throws UuriSerializationException {
    ...
  }
  /**
   * @returns The given URI's long form string representation.
   * @throws UuriSerializationException if the URI cannot be serialized using the
   *                     long form. The exception may contain details regarding
   *                     the violated constraint(s).
   */
  public static String serializeLongFormUri(UUri: uri) throws UuriSerializationException {
    ...
  }
}
----

Alternatively, the `UUri` class might provide corresponding methods.

[source,java]
----

> public class UUri {

Reviewer: UUri is defined in proto and the classes are generated to which we cannot extend. I recommend we do not show coding examples in this spec but state that serializers and deserializers and validators are required.

Author: ok, so simply remove the code examples?

  }
  pub fn validate_rpc_method(uri: &UUri) -> Result<(), UuriValidationError> {
    if is_rpc_method(uri) {
      Ok(())
    } else {
      Err(UuriValidationError::new("not an RPC Method URI"))
    }
  }
}
----

or implement the functions on the `UUri` struct

[source,rust]
----
impl UUri {
  pub fn is_rpc_method(&self) -> bool {
    ...
  }
  pub fn validate_rpc_method(&self) -> Result<(), UuriValidationError> {
    if self.is_rpc_method() {
      Ok(())
    } else {
      Err(UuriValidationError::new("not an RPC Method URI"))
    }
  }
}
----
--

== Pattern Matching

A UUri can be used to define a _pattern_ that other UUris can then be matched against.

.REQ~PATTERN-MATCHING
[#req~pattern-matching~1]
--
A _candidate_ UUri is said to _match_ a particular _pattern_ UUri if all of the candidate UUri's components match the pattern UUri's corresponding components according to the rules defined by the predicates below.

[source,ocl]
----
context UAuthority
def: matches(candidate : UAuthority) : Boolean =
  ( self.name->notEmpty() implies candidate.name->notEmpty()
    -- authority.name is case-insensitive per RFC3986
    and self.name.toLower() = candidate.name.toLower() )
  and

>   ( self.ip->notEmpty() implies candidate.ip->notEmpty() and self.ip = candidate.ip )

Reviewer: we can get rid of the ip and id

  ( self.name->notEmpty() implies candidate.name->notEmpty() and self.name = candidate.name )
  and
  ( self.id->notEmpty() implies candidate.id->notEmpty() and self.id = candidate.id )
  and
  ( self.version_major->notEmpty() implies candidate.version_major->notEmpty()
    and self.version_major = candidate.version_major )
  and
  ( self.version_minor->notEmpty() implies candidate.version_minor->notEmpty()
    and self.version_minor = candidate.version_minor )

context UResource
def: matches(candidate : UResource) : Boolean =
  ( self.name->notEmpty() implies candidate.name->notEmpty() and self.name = candidate.name )
  and
  ( self.instance->notEmpty() implies candidate.instance->notEmpty()
    and self.instance = candidate.instance )
  and
  ( self.message->notEmpty() implies candidate.message->notEmpty()
    and self.message = candidate.message )
  and
  ( self.id->notEmpty() implies candidate.id->notEmpty() and self.id = candidate.id )

context UUri
def: matches(candidate : UUri) : Boolean =
  self.authority.matches(candidate.authority)
  and
  self.entity.matches(candidate.entity)
  and
  self.resource.matches(candidate.resource)
----

[NOTE,#matching-local-uuris]
====

> An empty authority pattern matches _any_ authority according to the `UAuthority::matches` predicate. In particular, it is not possible to use an empty authority pattern to filter out _non-local_ candidate UUris.

Reviewer: I think I get it, you're saying (had to scratch my head multiple times).

//vcu.other.device/usubscription/3/SubscriptionChange#Update
----

NOTE: It seems counter-intuitive that the first URI does not match the pattern. Please refer to <<REQ~PATTERN-MATCHING>> for details on how to match _local_ UUris.

==== Matching on UResource

The following pattern UUri

[example]
----
UUri {
  UAuthority {},
  UEntity {
    name: "core.usubscription",
  },
  UResource {
    name: "rpc"
  }
}
----

will match all of the following (long form) URIs:
[example]
----
//other-vcu.my-vehicle/core.usubscription/3/rpc.Subscribe
//mcu.my-vehicle/core.usubscription/2/rpc.RegisterForNotifications
----

But not these:
[example]
----
//vcu.other.device/core.usubscription/1/SubscriptionChange#Update
//vcu.other.device/usubscription/3/rpc.Subscribe
----

== Best Practices

When a UUri is not resolved, i.e. only contains names, the only way to resolve the UUri is to lookup in uDiscovery the missing information. UEntity name and numbers are declared in the uService's protos by adding the service metadata attributes using https://protobuf.dev/programming-guides/proto2/#options[Protobuf Options]. This information for the uE is then exposed through the uService generated stubs.
 

> UResource portion of the UUri is also encoded in the uService protos and can be fetched by application developers using https://protobuf.dev/reference/csharp/api-docs/class/google/protobuf/reflection/message-descriptor[MessageDescriptors]. Application developers are then able to easily construct resolved UUris using the above mentioned generated code. link:../up-l2/README.adoc[Dispatchers] however do not have the name-to-number mapping metadata as they are only provided either serialized Long or Micro Uris so if they have to send a Uri in a different format, they have no choice but to dynamically lookup (resolve) the Uri. 

Reviewer: ```suggestion
UResource portion of the UUri is also encoded in the uService protos and can be fetched by application developers using https://protobuf.dev/reference/csharp/api-docs/class/google/protobuf/reflection/message-descriptor[MessageDescriptors]. Application developers are then able to easily construct resolved UUris using the above mentioned generated code. link:../up-l2/README.adoc[Dispatchers] however do not have the name-to-number mapping metadata as they are only provided serialized Uris so if they have to send a Uri in a different format, they have no choice but to dynamically lookup (resolve) the Uri. 
```

> IMPORTANT: If a uE has a specific UUri format requirements (Long vs Micro), they *MUST* document them in their respective APIs specifications. Please see link:../up-l3/README.adoc#_core_platform_ues[Core uProtocol uEs] for more specific requirements for uProtocol core services.

Reviewer: ```suggestion
IMPORTANT: If a uE has a specific UUri format requirements (UUri object, `string` Long/Short, or `bytes` format  Micro), they *MUST* document them in their respective APIs specifications. Please see link:../up-l3/README.adoc#_core_platform_ues[Core uProtocol uEs] for more specific requirements for uProtocol core services.
```
